leshan-server-core/src/main/java/org/eclipse/leshan/server/bootstrap/BootstrapHandler.java

```diff
@@ -156,40 +165,48 @@ private void sendServers(final String endpoint, final Identity destination, fina
+            send(bsSession, new BootstrapWriteRequest(path, serverInstance, ContentFormat.TLV),
                     new ResponseCallback<BootstrapWriteResponse>() {
                         @Override
                         public void onResponse(BootstrapWriteResponse response) {
-                            LOG.debug("Bootstrap write {} return code {}", endpoint, response.getCode());
+                            LOG.debug("Bootstrap write {} return code {}", bsSession.getEndpoint(), response.getCode());
                             // recursive call until toSend is empty
-                            sendServers(endpoint, destination, cfg, toSend);
+                            sendServers(bsSession, cfg, toSend);
                         }
                     }, new ErrorCallback() {
                         @Override
                         public void onError(Exception e) {
                             // TODO Handle error on bootstrap
                             LOG.warn(String.format("Error pending bootstrap write of server instance %s on %s",
-                                    serverInstance, endpoint), e);
+                                    serverInstance, bsSession.getEndpoint()), e);
                         }
                     });
         } else {
-            requestSender.send(endpoint, destination.getPeerAddress(), destination.isSecure(),
-                    new BootstrapFinishRequest(), new ResponseCallback<BootstrapFinishResponse>() {
-                        @Override
-                        public void onResponse(BootstrapFinishResponse response) {
-                            LOG.debug("Bootstrap Finished {} return code {}", endpoint, response.getCode());
-                        }
-                    }, new ErrorCallback() {
-                        @Override
-                        public void onError(Exception e) {
-                            // TODO Handle error on bootstrap
-                            LOG.warn(String.format("Error pending bootstrap finished on %s", endpoint), e);
-                        }
-                    });
+            send(bsSession, new BootstrapFinishRequest(), new ResponseCallback<BootstrapFinishResponse>() {
+                @Override
+                public void onResponse(BootstrapFinishResponse response) {
+                    LOG.debug("Bootstrap Finished {} return code {}", bsSession.getEndpoint(), response.getCode());
+                    // TODO(pht) check error code ?
+                    if (response.isSuccess()) {
+                        bsSessionManager.end(bsSession);
+                    }
+                }
+            }, new ErrorCallback() {
+                @Override
+                public void onError(Exception e) {
+                    // TODO Handle error on bootstrap
+                    LOG.warn(String.format("Error pending bootstrap finished on %s", bsSession.getEndpoint()), e);
```

Comment:
I want to separate a proper ending from an error. WDTA three methods `begin` / `end` / `failed` ?